build-tools/md/main.go

```diff
@@ -0,0 +1,205 @@
+// Command md is a markdown processing tool
+//
+// md generates evy frontend code
+package main
+
+import (
+	"bytes"
+	"embed"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+	"text/template"
+
+	"github.com/alecthomas/kong"
+	"rsc.io/markdown"
+)
+
+//go:embed tmpl/*
+var tmplFS embed.FS
+
+type app struct {
+	SrcDir  string `arg:"" type:"existingdir" help:"source directory" placeholder:"SRC-DIR"`
+	DestDir string `arg:"" help:"target directory" placeholder:"TARGET-DIR"`
+}
+
+func main() {
+	kctx := kong.Parse(&app{})
+	kctx.FatalIfErrorf(kctx.Run())
+}
+
+func (a *app) Run() error {
+	mdFiles, err := a.copy()
+	if err != nil {
+		return err
+	}
+	return a.genHTMLFiles(mdFiles)
+}
+
+// Copy the contents of the `src` directory to the `dest` directory.
+// Skip over *.md files as we will generate *.html files from them.
+func (a *app) copy() ([]string, error) {
+	mdFiles := []string{}
+	srcFS := os.DirFS(a.SrcDir)
+	err := fs.WalkDir(srcFS, ".", func(filename string, d fs.DirEntry, err error) error {
+		if err != nil {
+			// Errors from WalkDir do not include `src` in the path making
+			// the error messages not useful. Add src back in.
+			var pe *fs.PathError
+			if errors.As(err, &pe) {
+				pe.Path = filepath.Join(a.SrcDir, pe.Path)
+				return pe
+			}
+			return err
+		}
+		srcfile := filepath.Join(a.SrcDir, filename)
+		destfile := filepath.Join(a.DestDir, filename)
+
+		if d.IsDir() {
+			return os.MkdirAll(destfile, 0o777) //nolint:gosec
+		}
+
+		if filepath.Ext(filename) == ".md" {
+			mdFiles = append(mdFiles, filename)
+			return nil
+		}
+		sf, err := os.Open(srcfile)
+		if err != nil {
+			return err
+		}
+		defer sf.Close() //nolint:errcheck // don't care about close failing on read-only files
+		df, err := os.Create(destfile)
+		if err != nil {
+			df.Close() //nolint:errcheck,gosec // we're returning the more important error
+			return err
+		}
+		if _, err := io.Copy(df, sf); err != nil {
+			df.Close() //nolint:errcheck,gosec // we're returning the more important error
+			return err
+		}
+		return df.Close()
+	})
+	if err != nil {
+		return nil, err
+	}
+	return mdFiles, nil
+}
+
+func (a *app) genHTMLFiles(mdFiles []string) error {
+	for _, mdf := range mdFiles {
+		mdFile := filepath.Join(a.SrcDir, mdf)
+		htmlFile := filepath.Join(a.DestDir, htmlFilename(mdf))
+		root := toRoot(mdf)
+		err := genHTMLFile(mdFile, htmlFile, root)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+var tmpl = template.Must(template.ParseFS(tmplFS, "tmpl/docs.html.tmpl"))
+
+type tmplData struct {
+	Root  string
+	Title string
+	Main  string
```

Comment:
Perhaps `Body` or `Content`? I'm looking at this going "what is main? I don't understand"

Reply:
There will be table of content in a separate HTML tag.
Also it is currently rendered to a `<main> </main>`

Reply:
ok renamed to `Content`.